Feature test ability in c10

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131248

Differential Revision: [D59978687](https://our.internmc.facebook.com/intern/diff/D59978687)

zdevito committed:
Feature test ability in c10

[ghstack-poisoned]

zdevito committed:
Update on "Feature test ability in c10"




[ghstack-poisoned]

zdevito committed:
Update on "Feature test ability in c10"


Differential Revision: [D59978687](https://our.internmc.facebook.com/intern/diff/D59978687)

[ghstack-poisoned]

zdevito commented:
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59978687).

zdevito committed:
Update on "Feature test ability in c10"


Differential Revision: [D59978687](https://our.internmc.facebook.com/intern/diff/D59978687)

[ghstack-poisoned]

zdevito committed:
Update on "Feature test ability in c10"


Differential Revision: [D59978687](https://our.internmc.facebook.com/intern/diff/D59978687)

[ghstack-poisoned]

zdevito committed:
Update on "Feature test ability in c10"


Differential Revision: [D59978687](https://our.internmc.facebook.com/intern/diff/D59978687)

[ghstack-poisoned]

d4l3k commented on torch/csrc/Module.cpp:
```diff
@@ -1546,6 +1575,7 @@ static PyMethodDef TorchMethods[] = { // NOLINT
      (PyCFunction)(void (*)())THPModule_has_torch_function_variadic,
      METH_FASTCALL,
      nullptr},
+    {"_c10_justknobs_check", THPModule_c10_justknobs_check, METH_VARARGS, nullptr},
```
Can we update https://github.com/pytorch/pytorch/blob/605dfd8fb4672beef93b385ff91df5f349f2d394/torch/_utils_internal.py#L154 to use this?

zdevito replied:
Ill update it, but I'd like to stack a PR that does because it will be that PR which breaks stuff while this one is pretty safe to land with the only use being the test.

zdevito commented on c10/util/Knobs.cpp:
```diff
@@ -0,0 +1,40 @@
+bool justknobs_check(const char* the_namespace, const char* feature_name, bool default_value) {
+  return resolver(the_namespace, feature_name, default_value);
+}
+bool justknobs_check_default(const char* the_namespace, const char* feature_name, bool default_value) {
```
The intent is to allow the distributor to choose how to do feature rollouts. fbcode -> via jk, oss -> via environments. I don't think having a mix makes sense because it can get complicated to figure out which applied after the fact.

zdevito commented on c10/util/Knobs.h:
```diff
@@ -0,0 +1,47 @@
+// There are no default values by design.
+
+// Default: true
+bool C10_EXPORT justknobs_check(const char* the_namespace, const char* feature_name, bool default_value);
```
Colin wanted to name it the same as the python one, so I named it that way.

zdevito commented on c10/util/Knobs.h:
```diff
@@ -0,0 +1,47 @@
+// There are no default values by design.
+
+// Default: true
+bool C10_EXPORT justknobs_check(const char* the_namespace, const char* feature_name, bool default_value);
```
actually, this turns out annoying because we are always passing string literals, but if we turn them into string_views, then the implementation has to copy the string because it isn't null terminated.

zdevito committed:
Update on "Feature test ability in c10"


Differential Revision: [D59978687](https://our.internmc.facebook.com/intern/diff/D59978687)

[ghstack-poisoned]

zdevito committed:
Update on "Feature test ability in c10"


Differential Revision: [D59978687](https://our.internmc.facebook.com/intern/diff/D59978687)

[ghstack-poisoned]
